File: Programmers/Level2/메뉴리뉴얼/메뉴리뉴얼.py
from itertools import combinations
from collections import defaultdict

def solution(orders, course):
    answer = []
    dic = defaultdict(int)
    
    for c in course:
        temp = []
        for o in orders:
            combi = combinations(sorted(o), c)
            temp += combi

        for t in temp:
            dic[t] += 1
            
        answer += [k for k, v in dic.items() if max(dic.values()) == v and v > 1]
        dic.clear()
        
    return [''.join(a) for a in sorted(answer)]

# 다른 풀이
from collections import Counter
from itertools import combinations

def solution(orders, course):
    answer = []

    for c in course:
        combi = []
        for o in orders:
            combi += combinations(sorted(o), c)
        
        most_ordered = Counter(combi).most_common()
        answer += [ k for k, v in most_ordered if v > 1 and v == most_ordered[0][1] ]
    
    return [ ''.join(v) for v in sorted(result) ]

File: Programmers/Level2/메뉴리뉴얼/메뉴리뉴얼.js
function solution(orders, course) {
    const answer = [];

    for (let c of course) {
        const map = new Map();
        for (let o of orders) {
            let combination = getCombinations(o.split("").sort(), c);

            for (let combi of combination) {
                const comb = combi.join("");
                map.set(comb, (map.get(comb) || 0) + 1);
            }
        }

        map.forEach((value, key, _) => {
            if (value === Math.max(...map.values()) && value >= 2)
                answer.push(key);
        });
    }

    return answer.sort();
}

const getCombinations = function (arr, selectNumber) {
    const results = [];
    if (selectNumber === 1) return arr.map((el) => [el]);

    arr.forEach((fixed, index, origin) => {
        const rest = origin.slice(index + 1);
        const combinations = getCombinations(rest, selectNumber - 1);
        const attached = combinations.map((el) => [fixed, ...el]);
        results.push(...attached);
    });
    return results;
};
